File: 정종길/프로그래머스/카카오 3회독/2023 KAKAO BLIND RECRUITMENT/(다시 풀기) 표 병합.java
import java.util.*;

class Solution {

    String[][] arrs; 
    int[] parents;      
    List<String> answer; 
    
    public List<String> solution(String[] commands) {
        answer = new ArrayList<>();
        arrs = new String[51][51];
        parents = new int[2501];

        for (int i = 1; i <= 2500; i++) {
            parents[i] = i;
        }

        for (String command : commands) {
            String[] li = command.split(" ");
            if (li[0].equals("UPDATE")) {
                if (li.length == 4) {
                    update_rc(Integer.parseInt(li[1]), Integer.parseInt(li[2]), li[3]);
                } else {
                    update_v(li[1], li[2]);
                }
            } else if (li[0].equals("MERGE")) {
                merge(Integer.parseInt(li[1]), Integer.parseInt(li[2]),
                      Integer.parseInt(li[3]), Integer.parseInt(li[4]));
            } else if (li[0].equals("UNMERGE")) {
                unmerge(Integer.parseInt(li[1]), Integer.parseInt(li[2]));
            } else {
                pprint(Integer.parseInt(li[1]), Integer.parseInt(li[2]));
            }
        }

        return answer;
    }

    private int rcToIndex(int r, int c) {
        return (r - 1) * 50 + c;
    }

    private int[] indexToRC(int idx) {
        int r = (idx - 1) / 50 + 1;
        int c = (idx - 1) % 50 + 1;
        return new int[] { r, c };
    }

    private int find_parents(int idx) {
        if (parents[idx] != idx) {
            parents[idx] = find_parents(parents[idx]);
        }
        return parents[idx];
    }

    private void merge(int r1, int c1, int r2, int c2) {
        int idx1 = rcToIndex(r1, c1);
        int idx2 = rcToIndex(r2, c2);

        int rep1 = find_parents(idx1);
        int rep2 = find_parents(idx2);

        if (rep1 == rep2) {
            return;
        }

        int[] rc1 = indexToRC(rep1);
        int[] rc2 = indexToRC(rep2);

        parents[rep2] = rep1;

        String value1 = arrs[rc1[0]][rc1[1]];
        String value2 = arrs[rc2[0]][rc2[1]];

        if (value1 != null && value2 != null) {
            arrs[rc2[0]][rc2[1]] = null;
        } else if (value1 == null && value2 != null) {
            arrs[rc1[0]][rc1[1]] = value2;
            arrs[rc2[0]][rc2[1]] = null;
        } else if (value1 != null && value2 == null) {
            arrs[rc2[0]][rc2[1]] = null;
        }
    }

    private void unmerge(int r, int c) {
        int targetIdx = rcToIndex(r, c);

        for (int i = 1; i <= 2500; i++) {
            find_parents(i);
        }

        int rep = find_parents(targetIdx);
        int[] repRC = indexToRC(rep);
        String savedValue = arrs[repRC[0]][repRC[1]];


        for (int i = 1; i <= 2500; i++) {
            if (find_parents(i) == rep) {
                parents[i] = i;
                if (i != targetIdx) {
                    int[] cellRC = indexToRC(i);
                    arrs[cellRC[0]][cellRC[1]] = null;
                }
            }
        }

        arrs[repRC[0]][repRC[1]] = null;
        arrs[r][c] = savedValue;
    }

    private void update_rc(int r, int c, String v) {
        int idx = rcToIndex(r, c);
        int rep = find_parents(idx);
        int[] repRC = indexToRC(rep);
        arrs[repRC[0]][repRC[1]] = v;
    }

    private void update_v(String v1, String v2) {
        for (int i = 1; i <= 50; i++) {
            for (int j = 1; j <= 50; j++) {
                if (arrs[i][j] != null && arrs[i][j].equals(v1)) {
                    arrs[i][j] = v2;
                }
            }
        }
    }

    private void pprint(int r, int c) {
        int idx = rcToIndex(r, c);
        int rep = find_parents(idx);
        int[] repRC = indexToRC(rep);
        if (arrs[repRC[0]][repRC[1]] == null) {
            answer.add("EMPTY");
        } else {
            answer.add(arrs[repRC[0]][repRC[1]]);
        }
    }
}

File: 정종길/프로그래머스/카카오 3회독/2023 KAKAO BLIND RECRUITMENT/(다시 풀기) 표 병합.py
## 2회독 풀이로 참고하기 -> merge부분 제일 최적화 됨

def solution(commands):
    global excels, parents
    
    answer = []
    excels = [[0] * 51 for _ in range(51)]
    parents = [[(i, j) for j in range(51)] for i in range(51)]
    
    for command in commands:
        li = command.split()
        if li[0][:2] == "UP":
            if len(li) == 3:
                update3(li)
            else:
                update4(li)
                
        elif li[0][:2] == "ME":   
            merge(li)
            
        elif li[0][:2] == "UN":
            unmerge(li)
        else:
            answer.append(cprint(li))
    return answer

def find_parents(i, j):
    if parents[i][j] != (i, j):
        parents[i][j] = find_parents(parents[i][j][0], parents[i][j][1])
    return parents[i][j]
    
def update4(li):
    li[1], li[2] = int(li[1]), int(li[2])
    r, c = find_parents(li[1], li[2])
    excels[r][c] = li[3] 
    return

def update3(li):
    for r in range(51):
        for c in range(51):
            if excels[r][c] == li[1]:
                excels[r][c] = li[2]
    return

def merge(li):
    li[1], li[2], li[3], li[4] = int(li[1]), int(li[2]), int(li[3]), int(li[4])
    r1, c1 = find_parents(li[1], li[2])
    r2, c2 = find_parents(li[3], li[4])

    if (r1, c1) == (r2, c2):
        return
    
    if excels[r1][c1] and excels[r2][c2]:
        value = excels[r1][c1]
    else:
        if excels[r1][c1]:
            value = excels[r1][c1]
        else:
            value = excels[r2][c2]
            
    if r1 < r2:
        parents[r2][c2] = (r1, c1)
        excels[r1][c1] = value
        excels[r2][c2] = 0
    elif r2 < r1:
        parents[r1][c1] = (r2, c2)
        excels[r2][c2] = value
        excels[r1][c1] = 0
    else:
        if c1 < c2:
            parents[r2][c2] = (r1, c1)
            excels[r1][c1] = value
            excels[r2][c2] = 0
        else:
            parents[r1][c1] = (r2, c2)
            excels[r2][c2] = value
            excels[r1][c1] = 0
    return

def unmerge(li):
    for i in range(1, 51):
        for j in range(1, 51):
            find_parents(i, j)
    
    li[1], li[2] = int(li[1]), int(li[2])
    r, c = find_parents(li[1], li[2])
    value = 0
    
    if excels[r][c]:
        value = excels[r][c]
        excels[r][c] = 0
    
    for i in range(1, 51):
        for j in range(1, 51):
            if parents[i][j] == (r, c):
                parents[i][j] = (i, j)
    
    excels[li[1]][li[2]] = value
    return
    
def cprint(li):
    for i in range(1, 51):
        for j in range(1, 51):
            find_parents(i, j)
            
    li[1], li[2] = int(li[1]), int(li[2])
    r, c = find_parents(li[1], li[2])
    if not excels[r][c]:
        return "EMPTY"
    else:
        return excels[r][c]